vms/platformvm/validators/manager.go

```diff
@@ -170,56 +171,105 @@ func (m *manager) GetValidatorSet(ctx context.Context, height uint64, subnetID i
 		return nil, ErrMissingValidator
 	}
 
-	currentSubnetValidatorList := currentSubnetValidators.List()
-	vdrSet := make(map[ids.NodeID]*validators.GetValidatorOutput, len(currentSubnetValidatorList))
-	for _, vdr := range currentSubnetValidatorList {
-		primaryVdr, ok := currentPrimaryNetworkValidators.Get(vdr.NodeID)
+	currentSubnetValidatorList := currentSubnetValidatorSet.List()
+	currentSubnetValidators := make([]*maskedValidator, 0, len(currentSubnetValidatorList))
+	currentSubnetValidatorIndices := make(map[ids.NodeID]int)
+	for i, validator := range currentSubnetValidatorList {
+		primaryVdr, ok := currentPrimaryNetworkValidators.Get(validator.NodeID)
 		if !ok {
 			// This should never happen
-			return nil, fmt.Errorf("%w: %s", ErrMissingValidator, vdr.NodeID)
-		}
-		vdrSet[vdr.NodeID] = &validators.GetValidatorOutput{
-			NodeID:    vdr.NodeID,
-			PublicKey: primaryVdr.PublicKey,
-			Weight:    vdr.Weight,
+			return nil, fmt.Errorf("%w: %s", ErrMissingValidator, validator.NodeID)
 		}
+
+		currentSubnetValidators = append(currentSubnetValidators, &maskedValidator{
+			GetValidatorOutput: &validators.GetValidatorOutput{
+				NodeID:    validator.NodeID,
+				PublicKey: primaryVdr.PublicKey,
+				Weight:    validator.Weight,
+			},
+			masked: false,
+		})
+		currentSubnetValidatorIndices[validator.NodeID] = i
 	}
 
+	validatorsFromDiffs := make([]*maskedValidator, 0)
+	validatorsFromDiffsIndices := make(map[ids.NodeID]int)
 	for diffHeight := lastAcceptedHeight; diffHeight > height; diffHeight-- {
-		err := m.applyValidatorDiffs(vdrSet, subnetID, diffHeight)
-		if err != nil {
+		if validatorsFromDiffs, err = m.applyValidatorDiffs(
+			currentSubnetValidators,
+			currentSubnetValidatorIndices,
+			validatorsFromDiffs,
+			validatorsFromDiffsIndices,
+			subnetID,
+			diffHeight,
+		); err != nil {
 			return nil, err
 		}
 	}
 
+	// The current validator set is guaranteed to be sorted. Sort the set of
+	// validators calculated from applying diffs, so we can merge the two sorted
+	// lists.
```

Comment:
q: is it sorted by NodeID (as `validatorsFromDiffs` is below)? I don't where this order is guaranteed.

Reply:
as @abi87 points out, not sure what guaranteed the order for currentValidatorSet. Perhaps the ValidatorSet method on `state.State` should be replaced with a method that returns the validators in order?